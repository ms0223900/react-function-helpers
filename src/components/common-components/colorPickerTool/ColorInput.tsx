import React, { ChangeEvent } from 'react';
import { Box, TextField } from '@material-ui/core';

interface Props {
  onChange: (e: ChangeEvent<HTMLInputElement>) => any
  value: string
}
const ColorInput = ({
  onChange, value
}: Props) => {
  return (
    <TextField
      placeholder={'#ffffff'} 
      onChange={onChange}
      value={value} />
  );
};

export default ColorInput;