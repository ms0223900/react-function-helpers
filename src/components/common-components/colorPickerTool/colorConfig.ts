import { Color } from "@material-ui/core";


export const ITEM_WIDTH = 36;
export const PICKER_TOOL_WIDTH = ITEM_WIDTH * 4;
// 0-13
export type ColorShades = (keyof Color)[]
export const colorShades: ColorShades = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 'A100', 'A200', 'A400', 'A700'];
export const defaultColorShadeIndex = 5;