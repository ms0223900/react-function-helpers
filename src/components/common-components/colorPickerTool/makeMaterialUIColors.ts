
import { Color } from '@material-ui/core';
import * as Colors from '@material-ui/core/colors';

export const COLOR_NAME_REMOVED = 'common';

export const getMaterialColorNames = () => {
  const colorNames = Object.keys(Colors);
  const colorNamesRemoveCommon = colorNames.filter(color => color !== COLOR_NAME_REMOVED) as ColorKeys;
  return colorNamesRemoveCommon;
};

type ColorKeys = (keyof typeof Colors)[]
const makeAllColorsArr = () => {
  const colorNamesRemoveCommon = getMaterialColorNames();
  return colorNamesRemoveCommon.map(color => Colors[color]) as Color[];
};

const allColorsArr = makeAllColorsArr();

export default allColorsArr;