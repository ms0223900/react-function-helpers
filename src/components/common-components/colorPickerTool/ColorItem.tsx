import React from 'react';
import { Box, makeStyles } from '@material-ui/core';
import { CheckCircle } from '@material-ui/icons';
import { Callback } from 'all-common-types';
import { ITEM_WIDTH } from './colorConfig';

const useStyles = makeStyles(theme => ({
  root: {
    position: 'relative',
    cursor: 'pointer',
    width: ITEM_WIDTH,
    height: ITEM_WIDTH,
  },
  color: {
    width: '100%',
    height: '100%',
  },
  picked: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    margin: 'auto',
  }
}));

interface ColorItemProps {
  color: string,
  isPicked?: boolean,
  onClick?: Callback
}
export const ColorItem = ({ color, isPicked, onClick }: ColorItemProps) => {
  const classes = useStyles();
  return (
    <Box className={classes.root} onClick={onClick}>
      <Box 
        className={classes.color}
        style={{ backgroundColor: color }} />
      <Box 
        className={classes.picked} 
        display={'flex'} 
        alignItems={'center'} 
        justifyContent={'center'}
      >
        {!!isPicked && <CheckCircle />}
      </Box>
    </Box>
  );
};