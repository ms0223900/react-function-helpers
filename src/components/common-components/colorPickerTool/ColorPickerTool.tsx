import { Box, Color, makeStyles } from '@material-ui/core';
import React, { useCallback } from 'react';
import { PICKER_TOOL_WIDTH } from './colorConfig';
import { ColorItem,  } from './ColorItem';
import allColorsArr from './makeMaterialUIColors';

export interface GetPickedColorNowParams {
  colorsArr: Color[], 
  colorIndex: number | undefined, 
  colorShade: (keyof Color)
}
export const getPickedColorNow = ({colorsArr, colorIndex, colorShade}: {
  colorsArr: Color[], 
  colorIndex: number | undefined, 
  colorShade: (keyof Color)
}): string => {
  return typeof colorIndex === 'number' && colorsArr.length > 0 ?
    colorsArr[colorIndex][colorShade] : '';
};

const useStyles = makeStyles(() => ({
  root: {
    width: PICKER_TOOL_WIDTH,
  }
}));

export interface ColorPickerToolProps {
  colorShade: keyof Color
  pickedColorIndex: number | undefined
  pickFn: (index:number, color: string) => () => any
}
const ColorPickerTool = ({ colorShade, pickedColorIndex, pickFn }: ColorPickerToolProps) => {
  const classes = useStyles();
  const handlePick = useCallback((i: number, itemColor: string) => {
    return () => pickFn(i, itemColor)();
  }, [pickFn]);
  
  return (
    <Box className={classes.root} display={'flex'} flexWrap={'wrap'}>
      {allColorsArr.map((color, i) => {
        const itemColor = color[colorShade];
        const isPicked = i === pickedColorIndex;
        return (
          <ColorItem 
            key={i} 
            color={itemColor} 
            isPicked={isPicked} 
            onClick={handlePick(i, itemColor)} />
        );
      })}
    </Box>
  );
};


export default ColorPickerTool;