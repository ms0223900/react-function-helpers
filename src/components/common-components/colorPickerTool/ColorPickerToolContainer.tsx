import React, { useState, ChangeEvent, Ref, useImperativeHandle, forwardRef, useEffect, useCallback } from 'react';
import { Box, Typography } from '@material-ui/core';
import allColorsArr from './makeMaterialUIColors';
import ShaderSlider from './ShadeSlider';
import ColorInput from './ColorInput';
import { colorShades, defaultColorShadeIndex } from './colorConfig';
import ColorPickerTool, { getPickedColorNow, GetPickedColorNowParams } from './ColorPickerTool';
import PickedColors from './PickedColors';
import { PaletteKeys } from './types';
import { Callback } from 'all-common-types';

export interface ColorPickerToolContainerRef {
  getPickedColor: () => string
}
export interface ColorPickerToolContainerProps {
  label: PaletteKeys
}

export class Actions {
  static setColorAction = (params: GetPickedColorNowParams) => (dispatch: Callback) => {
    console.log('dispatch');
    const colorNow = getPickedColorNow(params);
    dispatch(colorNow);
  }
}


const ColorPickerToolContainer = (
  props: ColorPickerToolContainerProps, 
  ref: Ref<ColorPickerToolContainerRef>
) => {
  const { label } = props;
  const [pickedColor, setColor] = useState('#fff');
  const [pickedColorIndex, setColorIndex] = useState<number>();
  const [sliderVal, setSliderVal] = useState(defaultColorShadeIndex);
  useImperativeHandle(ref, () => ({
    getPickedColor: () => pickedColor
  }), [pickedColor]);
  
  const colorShade = colorShades[sliderVal];

  const handleInputColor = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    const color = e.target.value;
    setColor(color);
  }, []);
  const handlePickColor = useCallback((index:number, color: string) => {
    return () => {
      setColorIndex(index);
      setColor(color);
    };
  }, []);
  const handleChange = useCallback((e: ChangeEvent<{}>, val: any) => {
    setSliderVal(val);
  }, []);
  React.useEffect(() => {
    Actions.setColorAction({
      colorsArr: allColorsArr,
      colorIndex: pickedColorIndex,
      colorShade,
    })(setColor);
  }, [colorShade, pickedColorIndex]);

  return (
    <Box>
      <Typography variant={'h6'}>{label}</Typography>
      <PickedColors pickedColor={pickedColor} />
      <ColorInput onChange={handleInputColor} value={pickedColor} />
      <ColorPickerTool 
        colorShade={colorShade}
        pickedColorIndex={pickedColorIndex}
        pickFn={handlePickColor} />
      <ShaderSlider onChangeFn={handleChange} sliderVal={sliderVal} />
    </Box>
  );
};



export default forwardRef(ColorPickerToolContainer);