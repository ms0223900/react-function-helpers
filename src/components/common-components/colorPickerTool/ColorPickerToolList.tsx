import React, { MutableRefObject } from 'react';
import ColorPickerToolContainer, { ColorPickerToolContainerRef } from './ColorPickerToolContainer';
import { Button, makeStyles, Box } from '@material-ui/core';
import { PaletteKeys } from './types';

export type LabelsAndRefs<RefType> = {
  tabLabel: PaletteKeys,
  ref: MutableRefObject<RefType | null>
}[]
export interface ColorPickerToolListProps<RefType> {
  labelsAndRefs: LabelsAndRefs<RefType>
  handleSetThemeColorFn: () => any
}

const useStyles = makeStyles(() => ({
  root: {
    width: 600,
  }
}));

const ColorPickerToolList = ({
  labelsAndRefs, handleSetThemeColorFn
}: ColorPickerToolListProps<ColorPickerToolContainerRef>) => {

  const classes = useStyles();
  return (
    <Box className={classes.root}>
      <Box display={'flex'}>
        {labelsAndRefs.map((labelAndRef, i) => (
          <ColorPickerToolContainer
            key={i} 
            label={labelAndRef.tabLabel}
            ref={labelAndRef.ref} />
        ))}
      </Box>
      <Button 
        onClick={handleSetThemeColorFn} 
        variant={'contained'} 
        color={'primary'}
      >
        {'set theme color'}
      </Button>
    </Box>
  );
};

export default ColorPickerToolList;