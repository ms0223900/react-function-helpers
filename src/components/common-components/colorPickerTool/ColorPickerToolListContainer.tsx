import React, { useRef, MutableRefObject } from 'react';
import { ColorPickerToolContainerRef } from './ColorPickerToolContainer';
import { ThemeCustomizableColors } from './types';
import ColorPickerToolList from './ColorPickerToolList';
import { PaletteKeys } from './types';

export const getColorOfColorPickerTool = (ref: MutableRefObject<ColorPickerToolContainerRef | null>): string => (
  ref.current ? ref.current.getPickedColor() : ''
);

type LabelsAndRefs<Ref> = {
  tabLabel: PaletteKeys,
  ref: MutableRefObject<Ref>
}[]

export interface ColorPickerToolListContainerProps {
  setThemeColorsFn?: (colors: ThemeCustomizableColors) => any
}
const ColorPickerToolListContainer = ({
  setThemeColorsFn
}: ColorPickerToolListContainerProps) => {
  const PRIMARY_REF = useRef<ColorPickerToolContainerRef>(null);
  const SECONDARY_REF = useRef<ColorPickerToolContainerRef>(null);
  const labelsAndRefs = [
    { tabLabel: 'primary', ref: PRIMARY_REF, },
    { tabLabel: 'secondary', ref: SECONDARY_REF, },
  ] as LabelsAndRefs<ColorPickerToolContainerRef>;

  const handleSetThemeColor = () => {
    const themeColors: ThemeCustomizableColors = {
      primary: getColorOfColorPickerTool(labelsAndRefs[0].ref),
      secondary: getColorOfColorPickerTool(labelsAndRefs[1].ref),
    };
    setThemeColorsFn && setThemeColorsFn(themeColors);
  };

  return (
    <ColorPickerToolList
      labelsAndRefs={labelsAndRefs}
      handleSetThemeColorFn={handleSetThemeColor} />
  );
};

export default ColorPickerToolListContainer;