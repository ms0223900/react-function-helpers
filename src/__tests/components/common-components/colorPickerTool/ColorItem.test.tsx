import React from 'react';
import renderer from 'react-test-renderer';
import { ColorItem } from 'components/common-components/colorPickerTool/ColorItem';
import { shallow, mount } from 'enzyme';
import { CheckCircle } from '@material-ui/icons';

describe('test ColorItem', () => {
  const props = {
    color: '',
    isPicked: false
  };
  it('should render correctly', () => {
    const tree = renderer.create(
      <ColorItem {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have "CheckCircle" component if isPicked is true, vice versa.', () => {
    const wrapper = shallow(
      <ColorItem {...props} />
    );
    expect(wrapper.find(CheckCircle)).toHaveLength(0);

    const wrapperPicked = shallow(
      <ColorItem {...props} isPicked={true} />
    );
    expect(wrapperPicked.find(CheckCircle)).toHaveLength(1);
  });
});