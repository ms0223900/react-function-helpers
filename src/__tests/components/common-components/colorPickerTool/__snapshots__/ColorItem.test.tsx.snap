// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ColorItem should render correctly 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 makeStyles-root-1"
>
  <div
    className="MuiBox-root MuiBox-root-6 makeStyles-color-2"
    style={
      Object {
        "backgroundColor": "",
      }
    }
  />
  <div
    className="MuiBox-root MuiBox-root-7 makeStyles-picked-3"
  />
</div>
`;
